File: android/src/main/kotlin/unofficial/twilio/flutter/twilio_unofficial_programmable_video/PluginHandler.kt
package unofficial.twilio.flutter.twilio_unofficial_programmable_video

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.media.AudioAttributes
import android.media.AudioFocusRequest
import android.media.AudioManager
import android.os.Build
import androidx.annotation.NonNull
import com.twilio.video.AudioCodec
import com.twilio.video.CameraCapturer
import com.twilio.video.ConnectOptions
import com.twilio.video.G722Codec
import com.twilio.video.H264Codec
import com.twilio.video.IsacCodec
import com.twilio.video.LocalAudioTrack
import com.twilio.video.LocalParticipant
import com.twilio.video.LocalVideoTrack
import com.twilio.video.OpusCodec
import com.twilio.video.PcmaCodec
import com.twilio.video.PcmuCodec
import com.twilio.video.RemoteParticipant
import com.twilio.video.VideoCapturer
import com.twilio.video.VideoCodec
import com.twilio.video.Vp8Codec
import com.twilio.video.Vp9Codec
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import tvi.webrtc.voiceengine.WebRtcAudioUtils

class PluginHandler : MethodCallHandler, ActivityAware {
    private var previousAudioMode: Int = 0

    private var previousMicrophoneMute: Boolean = false

    private var audioFocusRequest: AudioFocusRequest? = null

    private var previousVolumeControlStream: Int = 0

    private var activity: Activity? = null

    private var applicationContext: Context

    private var myNoisyAudioStreamReceiver: BecomingNoisyReceiver

    private var audioManager: AudioManager

    private var headsetPlugIntent: IntentFilter

    constructor(applicationContext: Context) {
        this.applicationContext = applicationContext
        audioManager = applicationContext.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        myNoisyAudioStreamReceiver = BecomingNoisyReceiver(audioManager)
        headsetPlugIntent = IntentFilter(Intent.ACTION_HEADSET_PLUG)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        this.activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        this.activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        this.activity = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        this.activity = binding.activity
    }

    private val remoteParticipants: List<RemoteParticipant>?
        get() {
            return TwilioUnofficialProgrammableVideoPlugin.roomListener.room?.remoteParticipants?.toList()
        }

    fun getRemoteParticipant(sid: String?): RemoteParticipant? {
        return remoteParticipants?.first { it.sid == sid }
    }

    fun getLocalParticipant(): LocalParticipant? {
        return TwilioUnofficialProgrammableVideoPlugin.roomListener.room?.localParticipant
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        TwilioUnofficialProgrammableVideoPlugin.debug("TwilioUnofficialProgrammableVideoPlugin.onMethodCall => received ${call.method}")
        when (call.method) {
            "debug" -> debug(call, result)
            "connect" -> connect(call, result)
            "disconnect" -> disconnect(call, result)
            "setSpeakerphoneOn" -> setSpeakerphoneOn(call, result)
            "LocalAudioTrack#enable" -> localAudioTrackEnable(call, result)
            "LocalVideoTrack#enable" -> localVideoTrackEnable(call, result)
            "CameraCapturer#switchCamera" -> switchCamera(call, result)
            else -> result.notImplemented()
        }
    }

    private fun switchCamera(call: MethodCall, result: MethodChannel.Result) {
        TwilioUnofficialProgrammableVideoPlugin.debug("TwilioUnofficialProgrammableVideoPlugin.switchCamera => called")
        if (TwilioUnofficialProgrammableVideoPlugin.cameraCapturer != null) {
            TwilioUnofficialProgrammableVideoPlugin.cameraCapturer.switchCamera()
            return result.success(TwilioUnofficialProgrammableVideoPlugin.cameraCapturer.cameraSource.toString())
        }
        return result.error("NOT FOUND", "No CameraCapturer has been initialized yet natively", null)
    }

    private fun localVideoTrackEnable(call: MethodCall, result: MethodChannel.Result) {
        val localVideoTrackName = call.argument<String>("name")
        val localVideoTrackEnable = call.argument<Boolean>("enable")
        TwilioUnofficialProgrammableVideoPlugin.debug("TwilioUnofficialProgrammableVideoPlugin.localVideoTrackEnable => called for $localVideoTrackName, enable=$localVideoTrackEnable")
        if (localVideoTrackName != null && localVideoTrackEnable != null) {
            val localVideoTrack = TwilioUnofficialProgrammableVideoPlugin.roomListener.room?.localParticipant?.localVideoTracks?.firstOrNull { it.trackName == localVideoTrackName }
            if (localVideoTrack != null) {
                localVideoTrack.localVideoTrack.enable(localVideoTrackEnable)
                return result.success(true)
            }
            return result.error("NOT_FOUND", "No LocalVideoTrack found with the name '$localVideoTrackName'", null)
        }
        return result.error("MISSING_PARAMS", "The parameters 'name' and 'enable' were not given", null)
    }

    private fun localAudioTrackEnable(call: MethodCall, result: MethodChannel.Result) {
        val localAudioTrackName = call.argument<String>("name")
        val localAudioTrackEnable = call.argument<Boolean>("enable")
        TwilioUnofficialProgrammableVideoPlugin.debug("TwilioUnofficialProgrammableVideoPlugin.localAudioTrackEnable => called for $localAudioTrackName, enable=$localAudioTrackEnable")
        if (localAudioTrackName != null && localAudioTrackEnable != null) {
            val localAudioTrack = TwilioUnofficialProgrammableVideoPlugin.roomListener.room?.localParticipant?.localAudioTracks?.firstOrNull { it.trackName == localAudioTrackName }
            if (localAudioTrack != null) {
                localAudioTrack.localAudioTrack.enable(localAudioTrackEnable)
                return result.success(true)
            }
            return result.error("NOT_FOUND", "No LocalAudioTrack found with the name '$localAudioTrackName'", null)
        }
        return result.error("MISSING_PARAMS", "The parameters 'name' and 'enable' were not given", null)
    }

    private fun setSpeakerphoneOn(call: MethodCall, result: MethodChannel.Result) {
        val on = call.argument<Boolean>("on")
        if (on != null) {
            audioManager.isSpeakerphoneOn = on
            return result.success(on)
        }
        return result.error("MISSING_PARAMS", "The parameter 'value' was not given", null)
    }

    private fun disconnect(call: MethodCall, result: MethodChannel.Result) {
        TwilioUnofficialProgrammableVideoPlugin.debug("TwilioUnofficialProgrammableVideoPlugin.disconnect => called")
        TwilioUnofficialProgrammableVideoPlugin.roomListener.room?.localParticipant?.localVideoTracks?.forEach { it.localVideoTrack.release() }
        TwilioUnofficialProgrammableVideoPlugin.roomListener.room?.localParticipant?.localAudioTracks?.forEach { it.localAudioTrack.release() }
        TwilioUnofficialProgrammableVideoPlugin.roomListener.room?.localParticipant?.localDataTracks?.forEach { it.localDataTrack.release() }
        TwilioUnofficialProgrammableVideoPlugin.roomListener.room?.disconnect()
        setAudioFocus(false)
        result.success(true)
    }

    private fun connect(call: MethodCall, result: MethodChannel.Result) {
        TwilioUnofficialProgrammableVideoPlugin.debug("TwilioUnofficialProgrammableVideoPlugin.connect => called, Build.MODEL: '${Build.MODEL}'")
        if (TwilioUnofficialProgrammableVideoPlugin.HARDWARE_AEC_BLACKLIST.contains(Build.MODEL) && !WebRtcAudioUtils.useWebRtcBasedAcousticEchoCanceler()) {
            TwilioUnofficialProgrammableVideoPlugin.debug("WebRtcAudioUtils.setWebRtcBasedAcousticEchoCanceler => true")
            WebRtcAudioUtils.setWebRtcBasedAcousticEchoCanceler(true)
        }

        setAudioFocus(true)
        val optionsObj = call.argument<Map<String, Any>>("connectOptions")
        if (optionsObj != null) {
            try {
                val optionsBuilder = ConnectOptions.Builder(optionsObj["accessToken"] as String)

                // Set the room name if it has been passed.
                if (optionsObj["roomName"] != null) {
                    TwilioUnofficialProgrammableVideoPlugin.debug("TwilioUnofficialProgrammableVideoPlugin.connect => setting roomName to '${optionsObj["roomName"]}'")
                    optionsBuilder.roomName(optionsObj["roomName"] as String)
                }

                // Set the region if it has been passed.
                if (optionsObj["region"] != null) {
                    TwilioUnofficialProgrammableVideoPlugin.debug("TwilioUnofficialProgrammableVideoPlugin.connect => setting region to '${optionsObj["region"]}'")
                    optionsBuilder.region(optionsObj["region"] as String)
                }

                // Set the preferred audio codecs if it has been passed.
                if (optionsObj["preferredAudioCodecs"] != null) {
                    val preferredAudioCodecs = optionsObj["preferredAudioCodecs"] as Map<*, *>

                    val audioCodecs = ArrayList<AudioCodec>()
                    for ((audioCodec) in preferredAudioCodecs) {
                        when (audioCodec) {
                            IsacCodec.NAME -> audioCodecs.add(IsacCodec())
                            OpusCodec.NAME -> audioCodecs.add(OpusCodec())
                            PcmaCodec.NAME -> audioCodecs.add(PcmaCodec())
                            PcmuCodec.NAME -> audioCodecs.add(PcmuCodec())
                            G722Codec.NAME -> audioCodecs.add(G722Codec())
                            else -> audioCodecs.add(OpusCodec())
                        }
                    }
                    TwilioUnofficialProgrammableVideoPlugin.debug("TwilioUnofficialProgrammableVideoPlugin.connect => setting audioCodecs to '${audioCodecs.joinToString(", ")}'")
                    optionsBuilder.preferAudioCodecs(audioCodecs)
                }

                // Set the preferred video codecs if it has been passed.
                if (optionsObj["preferredVideoCodecs"] != null) {
                    val preferredVideoCodecs = optionsObj["preferredVideoCodecs"] as Map<*, *>

                    val videoCodecs = ArrayList<VideoCodec>()
                    for ((videoCodec) in preferredVideoCodecs) {
                        when (videoCodec) {
                            Vp8Codec.NAME -> videoCodecs.add(Vp8Codec()) // TODO(WLFN): It has an optional parameter, need to figure out for what: https://github.com/twilio/video-quickstart-android/blob/master/quickstartKotlin/src/main/java/com/twilio/video/quickstart/kotlin/VideoActivity.kt#L106
                            Vp9Codec.NAME -> videoCodecs.add(Vp9Codec())
                            H264Codec.NAME -> videoCodecs.add(H264Codec())
                            else -> videoCodecs.add(Vp8Codec())
                        }
                    }
                    TwilioUnofficialProgrammableVideoPlugin.debug("TwilioUnofficialProgrammableVideoPlugin.connect => setting videoCodecs to '${videoCodecs.joinToString(", ")}'")
                    optionsBuilder.preferVideoCodecs(videoCodecs)
                }

                // Set the local audio tracks if it has been passed.
                if (optionsObj["audioTracks"] != null) {
                    val audioTrackOptions = optionsObj["audioTracks"] as Map<*, *>

                    val audioTracks = ArrayList<LocalAudioTrack?>()
                    for ((audioTrack) in audioTrackOptions) {
                        audioTrack as Map<*, *> // Ensure right type.
                        audioTracks.add(LocalAudioTrack.create(this.applicationContext, audioTrack["enable"] as Boolean, audioTrack["name"] as String))
                    }
                    TwilioUnofficialProgrammableVideoPlugin.debug("TwilioUnofficialProgrammableVideoPlugin.connect => setting audioTracks to '${audioTracks.joinToString(", ")}'")
                    optionsBuilder.audioTracks(audioTracks)
                }

                // Set the local video tracks if it has been passed.
                if (optionsObj["videoTracks"] != null) {
                    val videoTrackOptions = optionsObj["videoTracks"] as Map<*, *>

                    val videoTracks = ArrayList<LocalVideoTrack?>()
                    for ((videoTrack) in videoTrackOptions) {
                        videoTrack as Map<*, *> // Ensure right type.
                        val videoCapturerMap = videoTrack["videoCapturer"] as Map<String, Any>

                        val videoCapturer: VideoCapturer = when (videoCapturerMap["type"] as String) {
                            "CameraCapturer" -> when (videoCapturerMap["cameraSource"] as String) {
                                "FRONT_CAMERA" -> CameraCapturer(this.applicationContext, CameraCapturer.CameraSource.FRONT_CAMERA)
                                "BACK_CAMERA" -> CameraCapturer(this.applicationContext, CameraCapturer.CameraSource.BACK_CAMERA)
                                else -> CameraCapturer(this.applicationContext, CameraCapturer.CameraSource.FRONT_CAMERA)
                            }
                            else -> CameraCapturer(this.applicationContext, CameraCapturer.CameraSource.FRONT_CAMERA)
                        }
                        if (videoCapturer is CameraCapturer) {
                            TwilioUnofficialProgrammableVideoPlugin.cameraCapturer = videoCapturer
                        }
                        videoTracks.add(LocalVideoTrack.create(this.applicationContext, videoTrack["enable"] as Boolean, videoCapturer, videoTrack["name"] as String))
                    }
                    TwilioUnofficialProgrammableVideoPlugin.debug("TwilioUnofficialProgrammableVideoPlugin.connect => setting videoTracks to '${videoTracks.joinToString(", ")}'")
                    optionsBuilder.videoTracks(videoTracks)
                }

                val roomId = 1 // Future preparation, for when we might want to support multiple rooms.
                TwilioUnofficialProgrammableVideoPlugin.roomListener = RoomListener(roomId, optionsBuilder.build())
                result.success(roomId)
            } catch (e: Exception) {
                result.error("INIT_ERROR", e.toString(), e)
            }
        } else {
            result.error("INIT_ERROR", "Missing ConnectOptions", null)
        }
    }

    private fun debug(call: MethodCall, result: MethodChannel.Result) {
        val enableNative = call.argument<Boolean>("native")
        if (enableNative != null) {
            TwilioUnofficialProgrammableVideoPlugin.nativeDebug = enableNative
            result.success(enableNative)
        } else {
            result.error("MISSING_PARAMS", "Missing 'native' parameter", null)
        }
    }

    private fun setAudioFocus(focus: Boolean) {
        if (focus) {
            previousAudioMode = audioManager.mode
            val volumeControlStream = this.activity?.volumeControlStream
            if (volumeControlStream != null) {
                previousVolumeControlStream = volumeControlStream
            }

            // Request audio focus
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                val playbackAttributes = AudioAttributes.Builder()
                        .setUsage(AudioAttributes.USAGE_VOICE_COMMUNICATION)
                        .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                        .build()
                audioFocusRequest = AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN_TRANSIENT)
                        .setAudioAttributes(playbackAttributes)
                        .setAcceptsDelayedFocusGain(true)
                        .setOnAudioFocusChangeListener { }
                        .build()
                audioManager.requestAudioFocus(audioFocusRequest)
            } else {
                audioManager.requestAudioFocus(null, AudioManager.STREAM_VOICE_CALL,
                        AudioManager.AUDIOFOCUS_GAIN_TRANSIENT)
            }
            /*
             * Use MODE_IN_COMMUNICATION as the default audio mode. It is required
             * to be in this mode when playout and/or recording starts for the best
             * possible VoIP performance. Some devices have difficulties with
             * speaker mode if this is not set.
             */
            audioManager.mode = AudioManager.MODE_IN_COMMUNICATION
            audioManager.setSpeakerphoneOn(!audioManager.isWiredHeadsetOn())
            applicationContext.registerReceiver(myNoisyAudioStreamReceiver, headsetPlugIntent)
            /*
             * Always disable microphone mute during a WebRTC call.
             */
            previousMicrophoneMute = audioManager.isMicrophoneMute

            audioManager.isMicrophoneMute = false
            this.activity?.volumeControlStream = AudioManager.STREAM_VOICE_CALL
        } else {
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
                audioManager.abandonAudioFocus(null)
            } else if (audioFocusRequest != null) {
                audioManager.abandonAudioFocusRequest(audioFocusRequest)
            }
            audioManager.setSpeakerphoneOn(false)
            audioManager.mode = previousAudioMode
            audioManager.isMicrophoneMute = previousMicrophoneMute
            this.activity?.volumeControlStream = previousVolumeControlStream
            try {
                applicationContext.unregisterReceiver(myNoisyAudioStreamReceiver)
            } catch (e: java.lang.Exception) {
                TwilioUnofficialProgrammableVideoPlugin.debug("${e.message}")
                e.printStackTrace()
            }
        }
    }
}

File: ios/Classes/PluginHandler.swift
import Flutter
import Foundation
import TwilioVideo

public class PluginHandler {
    let stillFrameRenderer: StillFrameRenderer = StillFrameRenderer()

    public func getRemoteParticipant(_ sid: String) -> RemoteParticipant? {
        return SwiftTwilioProgrammableVideoPlugin.roomListener?.room?.remoteParticipants.first(where: {$0.sid == sid})
    }

    public func getLocalParticipant() -> LocalParticipant? {
        return SwiftTwilioProgrammableVideoPlugin.roomListener?.room?.localParticipant
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.handle => received \(call.method)")
        switch call.method {
            case "debug":
                debug(call, result: result)
            case "connect":
                connect(call, result: result)
            case "disconnect":
                disconnect(call, result: result)
            case "setSpeakerphoneOn":
                setSpeakerphoneOn(call, result: result)
            case "getSpeakerphoneOn":
                getSpeakerphoneOn(result: result)
            case "takePhoto":
                takePhoto(result: result)
            case "LocalAudioTrack#enable":
                localAudioTrackEnable(call, result: result)
            case "LocalDataTrack#sendString":
                localDataTrackSendString(call, result: result)
            case "LocalDataTrack#sendByteBuffer":
                localDataTrackSendByteBuffer(call, result: result)
            case "LocalVideoTrack#enable":
                localVideoTrackEnable(call, result: result)
            case "CameraCapturer#switchCamera":
                switchCamera(call, result: result)
            default:
                result(FlutterMethodNotImplemented)
        }
    }

    func screenshotOfVideoStream(_ imageBuffer: CVImageBuffer?) -> Data {
        var ciImage: CIImage? = nil
        if let imageBuffer = imageBuffer {
            ciImage = CIImage(cvPixelBuffer: imageBuffer)
        }
        let temporaryContext = CIContext(options: nil)

        var videoImage: CGImage? = nil
        if let imageBuffer = imageBuffer, let ciImage = ciImage {
            videoImage = temporaryContext.createCGImage(
                ciImage,
                from: CGRect(x: 0, y: 0, width: CGFloat(CVPixelBufferGetWidth(imageBuffer)), height: CGFloat(CVPixelBufferGetHeight(imageBuffer))))
        }

        var image: UIImage? = nil
        if let videoImage = videoImage {
            image = UIImage(cgImage: videoImage)
        }
        if let imageData = image?.jpegData(compressionQuality: 1.0) {
            return imageData
        }

        return Data()
    }

    private func takePhoto(result: @escaping FlutterResult) {
        SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.takePhoto => called")
        if let frameToKeep = stillFrameRenderer.frameToKeep {
            return result(screenshotOfVideoStream(frameToKeep.imageBuffer))
        }
         return result(FlutterError(code: "NOT FOUND", message: "No frame data has been captured", details: nil))
    }

    private func switchCamera(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.switchCamera => called")
        if let cameraSource = SwiftTwilioProgrammableVideoPlugin.cameraSource {
            var captureDevice: AVCaptureDevice
            switch cameraSource.device?.position {
                case .back:
                    guard let realCaptureDevice = CameraSource.captureDevice(position: .front) else {
                        return result(RoomListener.videoSourceToDict(cameraSource))
                    }
                    captureDevice = realCaptureDevice
                default: // or .front
                    guard let realCaptureDevice = CameraSource.captureDevice(position: .back) else {
                        return result(RoomListener.videoSourceToDict(cameraSource))
                    }
                    captureDevice = realCaptureDevice
            }
            cameraSource.selectCaptureDevice(captureDevice, completion: { (_, _, error) in
                if let error = error {
                    return result(FlutterError(code: "\((error as NSError).code)", message: (error as NSError).description, details: nil))
                }
                return result(RoomListener.videoSourceToDict(cameraSource))
            })
        } else {
            return result(FlutterError(code: "NOT FOUND", message: "No CameraCapturer has been initialized yet natively", details: nil))
        }
    }

    private func localVideoTrackEnable(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'name' and 'enable' parameters", details: nil))
        }

        guard let localVideoTrackName = arguments["name"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'name' parameter", details: nil))
        }

        guard let localVideoTrackEnable = arguments["enable"] as? Bool else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'enable' parameter", details: nil))
        }

        SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.localVideoTrackEnable => called for \(localVideoTrackName), enable=\(localVideoTrackEnable)")

        let localVideoTrack = getLocalParticipant()?.localVideoTracks.first(where: {$0.trackName == localVideoTrackName})
        if let localVideoTrack = localVideoTrack {
            localVideoTrack.localTrack?.isEnabled = localVideoTrackEnable
            return result(true)
        }
        return result(FlutterError(code: "NOT_FOUND", message: "No LocalVideoTrack found with the name '\(localVideoTrackName)'", details: nil))
    }

    private func localAudioTrackEnable(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'name' and 'enable' parameters", details: nil))
        }

        guard let localAudioTrackName = arguments["name"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'name' parameter", details: nil))
        }

        guard let localAudioTrackEnable = arguments["enable"] as? Bool else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'enable' parameter", details: nil))
        }

        SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.localAudioTrackEnable => called for \(localAudioTrackName), enable=\(localAudioTrackEnable)")

        let localAudioTrack = getLocalParticipant()?.localAudioTracks.first(where: {$0.trackName == localAudioTrackName})
        if let localAudioTrack = localAudioTrack {
            localAudioTrack.localTrack?.isEnabled = localAudioTrackEnable
            return result(true)
        }
        return result(FlutterError(code: "NOT_FOUND", message: "No LocalAudioTrack found with the name '\(localAudioTrackName)'", details: nil))
    }

    private func localDataTrackSendString(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'name' and 'message' parameters", details: nil))
        }

        guard let localDataTrackName = arguments["name"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'name' parameter", details: nil))
        }

        guard let localDataTrackMessage = arguments["message"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'message' parameter", details: nil))
        }

        SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.localDataTrackSendString => called for \(localDataTrackName)")

        let localDataTrack = getLocalParticipant()?.localDataTracks.first(where: {$0.trackName == localDataTrackName})
        if let localDataTrack = localDataTrack {
            localDataTrack.localTrack?.send(localDataTrackMessage)
            return result(true)
        }
        return result(FlutterError(code: "NOT_FOUND", message: "No LocalDataTrack found with the name '\(localDataTrackName)'", details: nil))
    }

    private func localDataTrackSendByteBuffer(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'name' and 'message' parameters", details: nil))
        }

        guard let localDataTrackName = arguments["name"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'name' parameter", details: nil))
        }

        guard let localDataTrackMessage = arguments["message"] as? FlutterStandardTypedData else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'message' parameter", details: nil))
        }

        SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.localDataTrackSendString => called for \(localDataTrackName)")

        let localDataTrack = getLocalParticipant()?.localDataTracks.first(where: {$0.trackName == localDataTrackName})
        if let localDataTrack = localDataTrack {
            localDataTrack.localTrack?.send(localDataTrackMessage.data)
            return result(true)
        }
        return result(FlutterError(code: "NOT_FOUND", message: "No LocalDataTrack found with the name '\(localDataTrackName)'", details: nil))
    }

    private func setSpeakerphoneOn(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.setSpeakerphoneOn => called")
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'on' parameter", details: nil))
        }

        guard let on = arguments["on"] as? Bool else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'on' parameter", details: nil))
        }

        // If it is nil then this method is called before the connect method, so we can just set it here as well.
        if SwiftTwilioProgrammableVideoPlugin.audioDevice == nil {
            SwiftTwilioProgrammableVideoPlugin.audioDevice = DefaultAudioDevice()
            TwilioVideoSDK.audioDevice = SwiftTwilioProgrammableVideoPlugin.audioDevice!
        }

        do {
            DefaultAudioDevice.DefaultAVAudioSessionConfigurationBlock()

            let audioSession = AVAudioSession.sharedInstance()
            try on ? audioSession.setMode(.videoChat) : audioSession.setMode(.voiceChat)
            return result(on)
        } catch let error as NSError {
            return result(FlutterError(code: "\(error.code)", message: error.description, details: nil))
        }
    }

    private func getSpeakerphoneOn(result: @escaping FlutterResult) {
        let speakerPhoneOn = AVAudioSession.sharedInstance().mode == .videoChat
        SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.getSpeakerphoneOn => called \(speakerPhoneOn)")
        return result(speakerPhoneOn)
    }

    private func disconnect(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.disconnect => called")
        SwiftTwilioProgrammableVideoPlugin.roomListener?.room?.disconnect()

        if let camera = SwiftTwilioProgrammableVideoPlugin.cameraSource {
            camera.stopCapture()
            SwiftTwilioProgrammableVideoPlugin.cameraSource = nil
        }

        result(true)
    }

    // swiftlint:disable:next cyclomatic_complexity function_body_length
    private func connect(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.connect => called")
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'connectOpotions' parameter", details: nil))
        }

        guard let optionsObj = arguments["connectOptions"] as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'connectOptions' parameter", details: nil))
        }

        guard let accessToken = optionsObj["accessToken"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'accessToken' parameter", details: nil))
        }

        // Override the device before creating any Rooms or Tracks.
        if SwiftTwilioProgrammableVideoPlugin.audioDevice == nil {
            SwiftTwilioProgrammableVideoPlugin.audioDevice = DefaultAudioDevice()
            TwilioVideoSDK.audioDevice = SwiftTwilioProgrammableVideoPlugin.audioDevice!
        }

        let connectOptions = ConnectOptions(token: accessToken) { (builder) in
            // Set the room name if it has been passed.
            if let roomName = optionsObj["roomName"] as? String {
                SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.connect => setting roomName to '\(roomName)'")
                builder.roomName = roomName
            }

            // Set the region if it has been passed.
            if let region = optionsObj["region"] as? String {
                SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.connect => setting region to '\(region)'")
                builder.region = region
            }

            // Set the preferred audio codecs if it has been passed.
            if let preferredAudioCodecs = optionsObj["preferredAudioCodecs"] as? [String: String] {
                var audioCodecs: [AudioCodec] = []
                for (_, audioCodec) in preferredAudioCodecs {
                    switch audioCodec {
                        case "isac":
                            audioCodecs.append(IsacCodec())
                        case "PCMA":
                            audioCodecs.append(PcmaCodec())
                        case "PCMU":
                            audioCodecs.append(PcmuCodec())
                        case "G722":
                            audioCodecs.append(G722Codec())
                        default: // or opus
                            audioCodecs.append(OpusCodec())
                    }
                }
                SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.connect => setting preferredAudioCodecs to '\(audioCodecs)'")
                builder.preferredAudioCodecs = audioCodecs
            }

            // Set the preferred video codecs if it has been passed.
            if let preferredVideoCodecs = optionsObj["preferredVideoCodecs"] as? [String: String] {
                var videoCodecs: [VideoCodec] = []
                for (_, videoCodec) in preferredVideoCodecs {
                    switch videoCodec {
                        case "VP9":
                            videoCodecs.append(Vp9Codec())
                        case "H264":
                            videoCodecs.append(H264Codec())
                        default: // or VP8
                            videoCodecs.append(Vp8Codec())
                    }
                }
                SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.connect => setting preferredVideoCodecs to '\(videoCodecs)'")
                builder.preferredVideoCodecs = videoCodecs
            }

            // Set the local audio tracks if it has been passed.
            if let audioTrackOptions = optionsObj["audioTracks"] as? [AnyHashable: [String: Any]] {
                var audioTracks: [LocalAudioTrack] = []
                for (_, audioTrack) in audioTrackOptions {
                    let enable = audioTrack["enable"] as? Bool
                    let name = audioTrack["name"] as? String
                    audioTracks.append(LocalAudioTrack(options: nil, enabled: enable ?? true, name: name ?? nil)!)
                }
                SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.connect => setting audioTracks to '\(audioTracks)'")
                builder.audioTracks = audioTracks
            }

            // Set the local data tracks if it has been passed.
            if let dataTracksDict = optionsObj["dataTracks"] as? [AnyHashable: [String: Any]] {
                var dataTracks: [LocalDataTrack] = []
                for (_, dataTrack) in dataTracksDict {
                    if let dataTrackOptionsDict = dataTrack["dataTrackOptions"] as? [AnyHashable: Any] {
                        let dataTrackOptions = DataTrackOptions { (builder) in
                            if let ordered = dataTrackOptionsDict["ordered"] as? Bool {
                                builder.isOrdered = ordered
                            }
                            if let maxPacketLifeTime = dataTrackOptionsDict["maxPacketLifeTime"] as? Int32 {
                                builder.maxPacketLifeTime = maxPacketLifeTime
                            }
                            if let maxRetransmits = dataTrackOptionsDict["maxRetransmits"] as? Int32 {
                                builder.maxRetransmits = maxRetransmits
                            }
                            if let name = dataTrackOptionsDict["name"] as? String {
                                builder.name = name
                            }
                        }
                        dataTracks.append(LocalDataTrack(options: dataTrackOptions)!)
                    } else {
                        dataTracks.append(LocalDataTrack()!)
                    }
                }
                SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.connect => setting dataTracks to '\(dataTracks)'")
                builder.dataTracks = dataTracks
            }

            // Set the local video tracks if it has been passed.
            if let videoTrackOptions = optionsObj["videoTracks"] as? [AnyHashable: [String: Any]] {
                var videoTracks: [LocalVideoTrack] = []
                for (_, videoTrack) in videoTrackOptions {
                    let enable = videoTrack["enable"] as? Bool
                    let name = videoTrack["name"] as? String

                    let videoCapturer = videoTrack["videoCapturer"] as? [String: Any]
                    let videoSourceType = videoCapturer?["type"] as? String

                    switch videoSourceType {
                        default: // or CameraCapturer
                            let cameraSource = videoCapturer?["cameraSource"] as? String
                            let cameraDevice: AVCaptureDevice
                            switch cameraSource {
                                case "BACK_CAMERA":
                                    cameraDevice = CameraSource.captureDevice(position: .back)!
                                default: // or FRONT_CAMERA
                                    cameraDevice = CameraSource.captureDevice(position: .front)!
                            }
                            let videoSource = CameraSource()!
                            let localVideoTrack = LocalVideoTrack(source: videoSource, enabled: enable ?? true, name: name ?? nil)!

                            let videoFormat: VideoFormat = VideoFormat()
                            videoFormat.dimensions  = CMVideoDimensions(width:1920, height: 1080)
                            videoFormat.frameRate = 15

                            videoSource.startCapture(device: cameraDevice, format: videoFormat, completion: nil)
                            videoTracks.append(localVideoTrack)
                            SwiftTwilioProgrammableVideoPlugin.cameraSource = videoSource
                    }
                }
                SwiftTwilioProgrammableVideoPlugin.debug("PluginHandler.connect => setting videoTracks to '\(videoTracks)'")
                builder.videoTracks = videoTracks
            }
        }

        let roomId = 1
        SwiftTwilioProgrammableVideoPlugin.roomListener = RoomListener(roomId, connectOptions)
        result(roomId)
    }

    private func debug(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'native' parameter", details: nil))
        }

        guard let enableNative = arguments["native"] as? Bool else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'native' parameter", details: nil))
        }

        SwiftTwilioProgrammableVideoPlugin.nativeDebug = enableNative
        result(enableNative)
    }
}

class StillFrameRenderer: NSObject, VideoRenderer {
    var frameToKeep: VideoFrame?

    func renderFrame(_ frame: VideoFrame) {
        frameToKeep = frame
    }

    func updateVideoSize(_ videoSize: CMVideoDimensions, orientation: VideoOrientation) {
        //Do nothing
    }
}
